queue.go

```diff
@@ -386,40 +336,136 @@ func (c *client) registerReconnect(ctx context.Context) {
 				c.conn.Close()
 				c.conn = nil
 			}
-
-			channel := func() error {
-				var err error
-				c.channel, err = c.conn.Channel()
-				return errors.Wrap(err, "opening a new channel")
-			}
-			var err error
-			for {
-				time.Sleep(c.retryDelay)
-				if c.closed {
-					return
-				}
-				err = c.dial()
-				if err == nil {
-					err = channel()
-					if err == nil {
-						break
-					}
-				}
-			}
+			c.keepConnecting()
```

Comment:
Isn't there the chance that this blocks the entire goroutine?
Thinking of this being called in `NewClient`

Reply:
This is supposed to block the entire operation. This only time we get to this line is when we get disconnected from the broker, and we need to reconnect. The reason behind this method is to keep dialling until the broker can respond, hence `keepConnecting` name.

Regarding this to be called in the `NewClient` function, I think you are right. I propose we refactor this design later and abstract away these connections.